### ISSUE TITLE
[Question] Image Embeddings for ViT

### ISSUE THREAD

**Issue by hadminh** [xenova/transformers.js#321]
Is it possible to get image embeddings using  Xenova/vit-base-patch16-224-in21k model? We use feature_extractor to get embeddings for sentences. Can we use feature_extractor to get image embeddings?
```js
const model_id = "Xenova/vit-base-patch16-224-in21k";
const image = await RawImage.read("https://huggingface.co/datasets/Xenova/transformers.js-docs/resolve/main/football-match.jpg");
const classifier = await pipeline("image-classification", model_id);
const { image_embeddings } = await classifier.processor.feature_extractor(image);
```

**Comment by hadminh** [comment]
Thank you so much for a very detailed answer. 
Previously I used this method but I received an astronomically large dimension. Have you implemented a dimension reduction method or used one in js land?
```js
    const model_id = "Xenova/vit-base-patch16-224-in21k";
    // const model_id = "Xenova/clip-vit-base-patch16";
    // const processor = await AutoProcessor.from_pretrained(model_id);
    // const vision_model = await CLIPVisionModelWithProjection.from_pretrained(model_id, {
    //   quantized: false,
    // });

    const processor = await AutoProcessor.from_pretrained(model_id);
    const image = await RawImage.read("https://huggingface.co/datasets/Xenova/transformers.js-docs/resolve/main/football-match.jpg");
    const image_inputs = await processor(image);
    const vision_model = await ViTPreTrainedModel.from_pretrained(model_id, {
      quantized: false,
    });

// output:
// {
//   last_hidden_state: Tensor {
//     dims: [ 1, 197, 768 ],
//     type: 'float32',
//     data: Float32Array(151296) [
//         0.01596166007220745,  -0.12924863398075104,  -0.06474697589874268,
//         0.07320182025432587,  -0.24209485948085785,   -0.3889622092247009,
//       -0.014750899747014046, -0.009343968704342842,  -0.09137815982103348,
//       -0.016781888902187347,  -0.14880703389644623,   0.03444936126470566,
//        -0.24350781738758087,  -0.07449625432491302, -0.020611822605133057,
//           0.068354532122612,  -0.09426206350326538,   0.06079084426164627,
//         -0.2642066180706024,    0.1421915888786316,   0.03547587990760803,
//         0.05608220770955086,   0.18927136063575745,   0.03400838375091553,
//         -0.3091387152671814,   0.05749059095978737,  0.053838904947042465,
//        -0.25840312242507935,  -0.10178925096988678,  0.030108172446489334,
//       -0.031242091208696365,  -0.22251905500888824,   0.09145721048116684,
//       -0.057273026555776596,  -0.13317608833312988,  -0.15116924047470093,
//        -0.13443438708782196,  0.025702133774757385,   -0.0322958305478096,
//         -0.2062864899635315,   0.08286859095096588, 0.0004886187962256372,
//       -0.014200703240931034,   0.22494475543498993,  0.009350501000881195,
//       -0.054266154766082764,  -0.20810534060001373,  -0.15993162989616394,
//         -0.1460622400045395,   0.22473499178886414,   -0.2996913492679596,
//         -0.4333711862564087,  0.019104911014437675,  -0.07231348007917404,
//        -0.06940021365880966,  -0.06343454867601395,  -0.20415475964546204,
//          0.2380070835351944, -0.015462026931345463,     0.132570281624794,
//        -0.07365197688341141, -0.007615889422595501,  -0.11751672625541687,
//           0.091439388692379,   0.28438258171081543, -0.044682227075099945,
//         0.16475215554237366,   0.06064049154520035,  -0.10078104585409164,
//        -0.10937944054603577,   0.34963327646255493, -0.045526135712862015,
//         0.07839401066303253,   0.15061990916728973, -0.059041328728199005,
//        0.022792454808950424,  -0.40122637152671814,   0.00764797069132328,
//          -0.115696482360363,  0.015093778260052204,  0.053867120295763016,
//        -0.09852547198534012,   0.15397992730140686,   0.24015220999717712,
//         -0.1059536337852478,   0.23318903148174286,  -0.02166515588760376,
//        0.013454018160700798,   0.27011531591415405,  -0.14155690371990204,
//        -0.32532256841659546,   -0.1297968029975891,   0.24706953763961792,
//         -0.1759842187166214,   -0.1502475142478943,   -0.2843334674835205,
//        -0.20638823509216309,  0.013546660542488098, -0.013042806647717953,
//        -0.10837407410144806,
//       ... 151196 more items
//     ],
//     size: 151296
//   }
// }
```

**Comment by xenova** [comment]
The way I understand vision transformers, is that they try to treat images in the same was as normal transformers treat text, only replacing "tokens" with "patches" (see "[An Image is Worth 16x16 Words](https://arxiv.org/abs/2010.11929v2)"). So, in this case - and depending on your task - you could employ a similar technique used by other text embedding models: mean pooling across the patches.

As you've noted, the shape of your tensor is `[ 1, 197, 768 ]`, which you could perform mean pooling across the 2nd dimension (`197`) to obtain a tensor of shape `[1, 768]`. Intuitively, it should capture the "average idea of all patches". You should be able to use my `mean_pooling` function (see [here](https://github.com/xenova/transformers.js/blob/c367f9d68b809bbbf81049c808bf6d219d761d23/src/utils/tensor.js#L588)) from `./src/utils/tensor.js`: passing `last_hidden_state` as the first parameter, and setting `attention_mask` to a Tensor full of ones. Or you can just adapt the function to remove the attention mask portion.

You might fare better with other dimensionality reduction techniques, and I guess this is just something you'd need to experiment with. Let me know how it goes :)

**Comment by StoyanStAtanasov** [comment]
Hi @hadminh, what was the result. I want to try image similarity search with vision transformer's embeddings as well.
